Rules.py

```diff
@@ -211,18 +243,617 @@ def global_rules(world):
+                                        and (state.form('Deku') or state.form('Goron') or state.form('Zora')))
+    set_rule(world.get_location('Ocean Gossip Check'), lambda state: (state.has('Sonata of Awakening') or state.has('Goron Lullaby') or state.has('New Wave Bossa Nova'))
+                                        and (state.form('Deku') or state.form('Goron') or state.form('Zora')))
+    set_rule(world.get_location('Canyon Gossip Check'), lambda state: (state.has('Sonata of Awakening') or state.has('Goron Lullaby') or state.has('New Wave Bossa Nova'))
+                                        and (state.form('Deku') or state.form('Goron') or state.form('Zora')))
+    set_rule(world.get_location('4 Gossip Stone HP'),
+             lambda state: state.event('Swamp Gossip Check') and state.event('Mountain Gossip Check')
+                           and state.event('Ocean Gossip Check') and state.event('Canyon Gossip Check'))
+
+    set_rule(world.get_location('TF Peahat Grotto HP'), lambda state: state.form('Human') or state.form('Goron') or state.form('Zora'))
+    # how many ways to kill a peahat? lol
+    # for now I'm gonna assume deku can't, but everyone else can
+
+    # set_rule(world.get_location('TF Dodongo Grotto HP'), lambda state: True)
+    # ways to kill a dodongo? might be anyone actually
+
+    set_rule(world.get_entrance('To TF Business Scrub Grotto'), lambda state: state.event('Saw Scrub Fly In'))
+    set_rule(world.get_location('TF Business Scrub Grotto HP'), lambda state: state.has('Adult Wallet'))
+    # giving the moon's tear to the scrub is part of the connection requirement to get to the grotto, but once there
+    # you just need the adult's wallet
+
+    set_rule(world.get_location('TF Beehive Grotto HP'), lambda state: (state.can_blast() or state.form('Goron'))
+                                                                                  and state.can_use('Bow') and state.form('Zora'))
+    # boulder to get through, then shoot the bees, then drop to the bottom of the pool as zora
+    # actually shooting the bees might be possible as zora
+    # bees might just be more balloons lol
+    # todo: test this hp
+
+    # set_rule(world.get_location('TF Chest In The Grass'), lambda state: True)
+    # set_rule(world.get_location('TF Deku Baba Pit Chest'), lambda state: True)
+    set_rule(world.get_location('TF Chest On A Stump'), lambda state: state.can_use('Hookshot'))
+
+    # set_rule(world.get_location(''), lambda state: state)
+    # apparently if you look at some guys standing in the field through the telescope, there will be pits where they
+    # each were? dunno what's in them, but we might need to record those
+    # todo: test this
+
+    set_rule(world.get_entrance('To East Pillar Grotto'),
+             lambda state: state.can_use('Bottle') and state.has('Magic Beans') and state.has('Water') and (state.form('Human') or state.form('Zora')))
```

Comment:
isn't `has(water)` kinda redundant? if you have a bottle, you can get water, right?

Reply:
that's probly true, at the very least even in a super weird entrance shuffle, you can get water from south clock town
I'll leave it off because otherwise we'd have to document every pool of water lol